int-set/src/lib.rs

```diff
@@ -12,45 +12,92 @@
 //!
 //! When constructing a new IntSet from an existing lists of integer values the most efficient
 //! way to create the set is to initialize it from a sorted (ascending) iterator of the values.
+//!
+//! For a type to be stored in the IntSet it must implement the IntSetMember trait, and all
+//! unique values of that type must be able to be mapped to and from a unique u32 value.
+//! See the IntSetMember trait for more information.
 
 mod bitpage;
 mod bitset;
 
 use bitset::BitSet;
 use std::hash::Hash;
+use std::marker::PhantomData;
 use std::ops::RangeInclusive;
 
 /// A fast & efficient invertible ordered set for small (up to 32-bit) unsigned integer types.
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub struct IntSet<T>(Membership<T>);
+pub struct IntSet<T>(Membership, PhantomData<T>);
+
+/// Members of IntSet must implement this trait. Members of IntSet's must meet the following
+/// conditions to be used in an IntSet:
+///
+/// 1. Every possible unique value of T must be able map to and from a unique u32
+///    integer.
+///
+/// 2. The mapped u32 values must retain the same ordering as the values in T.
+///
+/// 3. all_values_iter() must iterate over all values in T in sorted order (ascending).
+///
+/// from_mapped() will only ever be called with u32 values that were generated by a
+/// call to to_mapped() with a valid value of T. So it doesn't need to correctly handle values
+/// that are outside the domain of T.
+pub trait IntSetMember<T> {
+    fn to_mapped(&self) -> u32;
+    fn from_mapped(member: InDomain) -> T;
+    fn all_values_sorted_iter() -> impl Iterator<Item = u32>;
+}
+
+/// Marks a mapped value as being in the domain of T for IntSetMember<T>. See IntSetMember
+/// for more information.
+pub struct InDomain {
```

Comment:
By convention if you're referencing types or identifiers that are used in the code you put them in backticks.

Also it's very easy to add links between items in your docs, and this has an additional benefit: during CI we check that these links actually reference things that exist, which means if you rename something your docs will break if you don't update them.

```suggestion
/// A value in the domain of `T` for some `IntSetMember<T>`.
///
/// See [`IntSetMember`] for more information.
pub struct InDomain {
```

(I didn't make the first `IntSetMember<T>` a link only because I'm not sure the autolinking works with generics? It probably does)